### ISSUE TITLE
Distributed TF hangs because of "CreateSession still waiting for response from worker...."

### ISSUE THREAD

**Issue by OscarDPan** [tensorflow/tensorflow#14662]
**UPDATE:** The first 2 posts are no longer appropriate to describe the issue. Please jump to my 3rd post.

Hi,

I followed the idea of this https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/benchmark_cnn.py#L1670 to implement a worker sync queue. Everything seemed to work fine except when I increased the size of the dataset or number of workers: all workers hang when they try to evaluate the sync op. Typical example of my code is as follow:

For each worker:
```
    def create_sync_queue_ops(self, op_prefix):
        """
        op_prefix: a string that denote where the sync is being used.
        """
        #Evenly distribute queues to all ps or workers
        device_name = "/job:ps/task:{}/cpu:0".format(self.sync_queue_counter % self.num_ps_nodes)
        self.sync_queue_counter += 1
        with tf.device(device_name):
            sync_queues = [tf.FIFOQueue(self.num_worker_nodes, [tf.bool], shapes=[[]], shared_name="{0}_{1}".format(op_prefix,i)) for i in xrange(self.num_worker_nodes)]
            token = tf.constant(False)
            queue_ops = []
            for i, q in enumerate(sync_queues):
                if i == self.worker_id:
                    queue_ops.append(tf.no_op())
                else:
                    queue_ops.append(q.enqueue(token))
            #Drain tokens off queue for this worker after enqueuing ops
            with tf.control_dependencies(queue_ops):
                wait_ops = sync_queues[self.worker_id].dequeue_many(len(sync_queues)-1)
            return wait_ops

(some graph definition...)
sess = tf.Session()
demo_sync_ops = self.create_sync_queue_ops("demo")
if self.is_chief_worker: #only execute by worker 0, other workers do nothing
    sess.run(tf.global_variables_initializer())
print "finishing message"
sess.run(demo_sync_ops)
```

I could **occasionally** see all workers hang after printing the "finishing message". 
My observation so far is that this only happened when dataset is huge or number of worker is big. e.g. 10+TB dataset with 300-500 workers.

I haven't been able to see why this occurred, not sure if it is a TF issue or some network bottleneck that I was not aware of. Any help would be much appreciated!

**Comment by OscarDPan** [comment]
Some updates:
I think the shared queue is inherently unstable, according to my experience and to some Googler who works on TF.
So I implemented the sync mechanism without using queues. But the hangs still happened, and it looked like this was due another problem than some servers just hung because the port it got assigned was taken away by others. (hung on the first sess.run())

So there were actually at least two things contributing to the hanging behavior.
Anyone who has more insights please feel free to shed light.

**Comment by drpngx** [comment]
Yes, we are moving to the dataset API. Would that work for you?

/CC @mrry

**Comment by OscarDPan** [comment]
Update:
Found the direct cause of the hanging problem that happened right after the first sess.run(). It's because the there are some workers indefinitely waiting for the connections to other servers to be established. ("CreateSession still waiting for response from worker...."). The amount of bad connections was random, and this did not happened in some particular hosts.

This happened quite often when I have large dataset (5TB+) and 300+ servers.

**Comment by OscarDPan** [comment]
@drpngx 
Thank you for your response. The dataset we have is some huge .npy files. So what I did is first loading them into memory and splitting it into batches, then I enqueued those batches into RandomShuffleQueue. Could you elaborate why you suggested dataset API?

**Comment by drpngx** [comment]
@OscarDPan the dataset API is easier to analyze than the queues, which can be hard to debug when they freeze. As long as you're OK with loading into memory then splitting, I think it should work just fine.

Could you try that instead?

**Comment by OscarDPan** [comment]
@drpngx were you suggesting me to use the dataset API to create a synchronization barrier? I am not sure if that helps with the problems I found as I mentioned in the post above.

And FYI I had no problem enqueuing/dequeuing training data to/from a RandomShuffleQueue. As to whether it is ok to use FIFOQueue to create sync barrier, I am not sure since there's the server connection problems hindering the analysis.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@OscarDPan I want to make sure it's not a queuing problem. Sorry, our time is limited and I am trying to isolate the problem and make 100% sure that it's not a queuing problem before we take a look.

**Comment by OscarDPan** [comment]
@drpngx 
I don't think it's a queuing problem. I updated the title that better describes the problem.

**Comment by mrry** [comment]
@OscarDPan Which version of TF are you using? 1.5.0rc0 (released today) includes an upgraded version of gRPC that fixes a deadlock bug that causes some servers to hang after a random period of time (although more quickly under heavier request load/at larger scale). Could you confirm whether the same problem still applies with the latest code?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by OscarDPan** [comment]
@mrry 
I was using TF1.3. There's been a lot of changes to my graph since I found the problem. I am still using 1.3 but now each worker don't see other workers and only see the necessary PS. This reduces the connections a lot and I haven't seen any hanging problem.

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by mrry** [comment]
Okay, I'm going to close this issue for now, since you seem to have found a workaround. I suspect TF 1.5 (and the upgraded gRPC in particular) will fix the original program. If the problem comes back and it can't be solved by upgrading, please feel free to reopen the issue.
